quantum/bb84.py

```diff
@@ -0,0 +1,133 @@
+#!/usr/bin/env python3
+"""
+Simulation of the Quantum Key Distribution (QKD) protocol called BB84,
+created by Charles Bennett and Gilles Brassard in 1984.
+
+BB84 is a key-distribution protocol that ensures secure key distribution
+using qubits instead of classical bits. The generated key is the result
+of simulating a quantum circuit. Our algorithm to construct the circuit
+is as follows:
+
+Alice generates two binary strings. One encodes the basis for each qubit:
+
+ - 0 -> {0,1} basis.
+ - 1 -> {+,-} basis.
+
+The other encodes the state:
+
+ - 0 -> |0> or |+>.
+ - 1 -> |1> or |->.
+
+Bob also generates a binary string and uses the same convention to choose
+a basis for measurement. Based on the following results, we follow the
+algorithm below:
+
+X|0> = |1>
+
+H|0> = |+>
+
+HX|0> = |->
+
+1. Whenever Alice wants to encode 1 in a qubit, she applies an
+X (NOT) gate to the qubit. To encode 0, no action is needed.
+
+2. Wherever she wants to encode it in the {+,-} basis, she applies
+an H (Hadamard) gate. No action is necessary to encode a qubit in
+the {0,1} basis.
+
+3. She then sends the qubits to Bob (symbolically represented in
+this circuit using wires).
+
+4. Bob measures the qubits according to his binary string for
+measurement. To measure a qubit in the {+,-} basis, he applies
+an H gate to the corresponding qubit and then performs a measurement.
+
+References:
+https://en.wikipedia.org/wiki/BB84
+https://qiskit.org/textbook/ch-algorithms/quantum-key-distribution.html
+"""
+import numpy as np
+import qiskit
+
+
+def bb84(key_len: int = 8, seed: int | None = None) -> str:
+    """
+    Performs the BB84 protocol using a key made of `key_len` bits.
+    The two parties in the key distribution are called Alice and Bob.
+    Args:
+        key_len (int): THe length of the generated key in bits. Default is 8.
+
+        seed (int, Optional): Seed for the random number generator.
+        Mostly used for testing. Default is None.
+
+    Returns:
+        key (str): The key generated using BB84 protocol.
```

Comment:
Let's not repeat the type that is already stated in the function signature.  The problem is that if the data type changes then the reader does not know which one to trust.  Type hints are probably more correct than comments because of the increasing use of tools like `mypy`, `pytype`, `pyre`, etc.  There is a meme among developers that "comments are lies".
.```suggestion
        key_len: The length of the generated key in bits. The default is 8.

        seed: Seed for the random number generator.
        Mostly used for testing. Default is None.

    Returns:
        key: The key generated using BB84 protocol.
```